lib/SPIRV/SPIRVRegularizeLLVM.cpp

```diff
@@ -85,6 +85,24 @@ class SPIRVRegularizeLLVM : public ModulePass {
   /// @spirv.llvm_memset_* and replace it with @llvm.memset.
   void lowerMemset(MemSetInst *MSI);
 
+  /// No SPIR-V counterpart for @llvm.fshl.* intrinsic. It will be lowered
+  /// to a newly generated @spirv.llvm_fshl_* function.
+  ///
+  /// Conceptually, FSHL:
+  /// 1. concatenates the ints, the first one being the more significant;
+  /// 2. performs a left shift-rotate on the resulting doubled-sized int;
+  /// 3. returns the most significant bits of the shift-rotate result,
+  ///    the number of bits being equal to the size of the original integers.
+  /// If FSHL operates on a vector type instead, the same operations are
+  /// performed for each set of corresponding vector elements.
+  ///
+  /// The actual implementation algorithm will be slightly different to speed
+  /// things up.
+  void lowerFunnelShiftLeft(IntrinsicInst *FSHLIntrinsic);
+  void buildFunnelShiftLeftFunc(Function *FSHLFunc);
```

Comment:
Why do need 2 functions? Do you anticipate other usages for `buildFunnelShiftLeftFunc`?

Reply:
At the moment of writing these, the logic of building FSHL involved 3 basic blocks, so I've chosen to split the functions for more concision. With `buildFunnelShiftLeftFunc` having become simpler, I personally still prefer this split: `lowerFunnelShiftLeft()` remains dedicated to function name production, function extraction and call replacement. Would it be preferable if I merged the functions instead?

Reply:
I don't have strong preference here. Up to you.